src/main/java/jenkins/branch/MultiBranchProjectFactory.java

```diff
@@ -46,21 +47,67 @@
-     * @return a new uninitialized multibranch project (do not configure its {@link MultiBranchProject#getSourcesList} or call {@link MultiBranchProject#onCreatedFromScratch}), or null if unrecognized
+     * @return true if recognized
      */
+    public boolean recognizes(@Nonnull ItemGroup<?> parent, @Nonnull String name, @Nonnull List<? extends SCMSource> scmSources, @Nonnull Map<String,Object> attributes, @Nonnull TaskListener listener) throws IOException, InterruptedException {
+        if (Util.isOverridden(MultiBranchProjectFactory.class, getClass(), "createProject", ItemGroup.class, String.class, List.class, Map.class, TaskListener.class)) {
+            return createProject(parent, name, scmSources, attributes, listener) != null;
+        } else {
+            throw new AbstractMethodError(getClass().getName() + " must override recognizes");
+        }
+    }
+
+    /**
+     * Creates a new multibranch project which matches {@link #recognizes}.
+     * @param parent a folder
+     * @param name a project name
+     * @param scmSources a set of SCM sources as added by {@link jenkins.scm.api.SCMSourceObserver.ProjectObserver#addSource}
+     * @param attributes a set of metadata attributes as added by {@link jenkins.scm.api.SCMSourceObserver.ProjectObserver#addAttribute}
+     * @param listener a way of reporting progress
+     * @return a new uninitialized multibranch project (do not configure its {@link MultiBranchProject#getSourcesList} or call {@link MultiBranchProject#onCreatedFromScratch})
+     */
+    @Nonnull
+    public MultiBranchProject<?,?> createNewProject(@Nonnull ItemGroup<?> parent, @Nonnull String name, @Nonnull List<? extends SCMSource> scmSources, @Nonnull Map<String,Object> attributes, @Nonnull TaskListener listener) throws IOException, InterruptedException {
+        if (Util.isOverridden(MultiBranchProjectFactory.class, getClass(), "createProject", ItemGroup.class, String.class, List.class, Map.class, TaskListener.class)) {
+            MultiBranchProject<?, ?> p = createProject(parent, name, scmSources, attributes, listener);
+            if (p == null) {
+                throw new IOException("recognized project " + name + " before, but now");
+            }
+            return p;
+        } else {
+            throw new AbstractMethodError(getClass().getName() + " must override createNewProject");
+        }
+    }
+
+    /**
+     * Updates an existing project which matches {@link #recognizes}.
+     * @param project an existing project, perhaps created by this factory, perhaps not
+     * @param attributes a set of metadata attributes as added by {@link jenkins.scm.api.SCMSourceObserver.ProjectObserver#addAttribute}
+     * @param listener a way of reporting progress
+     */
+    public void updateExistingProject(MultiBranchProject<?,?> project, @Nonnull Map<String,Object> attributes, @Nonnull TaskListener listener) throws IOException, InterruptedException {}
+
+    @Deprecated
```

Comment:
Not bothering with documentation, as the only implementation in production code was via `BySCMSourceCriteria` anyway.